phac-nml/irida-next
# Metadata Summary Update by Metadata::UpdateService

## What does this PR do and why?
As part of [issue 334](https://github.com/phac-nml/irida-next/issues/334), this PR implements `group` and `project` `metadata_summary` updating when metadata for a sample is changed via the `Samples::Metadata::UpdateServce`. 

When `sample_metadata` is updated by the `UpdateService`, the `project.namespace` and any `group` parents will have a count updated to track how many samples contain each metadata_field in their respective `update_summary`.

Note: updating `metadata_summary` by `project/group/sample` `transfer` will be implemented in a separate PR.
Note: Although not too clear in this PR, `subtract_from_metadata_summary` and `add_to_metadata_summary` were refactored to include the `update_by_one` argument in order for these functions to be usable by all `metadata_summary` updating (ie: transfers and deletions). See [PR 350](https://github.com/phac-nml/irida-next/pull/350/files) for context.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
Either create or utilize `groups/projects/samples` within `db:seed` with empty `metadata_summary` that would have a 'master' group, with separate descendent groups and/or projects.

I will use the following:
Bacillus 
     V
Bacillus cereus > Outbreak 2021 (var: `project1`) > Sample 10 (var: `sample1`)
             |
              ---------------- > Outbreak 2022 (var: `project2`) > Sample 10 (var: `sample2`)

1. In the rails console, assign the above variables `sample1`, `sample2`, `project1`, `project2`.
2. Assign a `user`  with permission to edit the above.
3. Test adding metadata
a) Assign `params1 = {'metadata' => {'metadatafield1' => 'value1'}}`
b) Run `Samples::Metadata::UpdateService.new(project1, sample1, user, params1).execute`
c) Check that Outbreak 2021, Bacillus cereus, and Bacillus now have an `metadata_summary = {'metadatafield1': 1}`. Outbreak 2022 `metadata_summary` should still be empty.
d) Run `Samples::Metadata::UpdateService.new(project2, sample2, user, params1).execute`
e) Check Bacillus cereus and Bacillus now have an `metadata_summary = {'metadatafield1': 2}`, and Outbreak 2021 and Outbreak 2022 `metadata_summary = {'metadatafield1': 1}` ` .

4. Test adding and deleting metadata
a) Assign `params2 = {'metadata' => {'metadatafield1' => '', 'metadatafield2' => 'value2'}}`
b) Run `Samples::Metadata::UpdateService.new(project1, sample1, user, params2).execute`
c) Check that Bacillus cereus, and Bacillus now have an `metadata_summary = {'metadatafield1': 1, 'metadatafield2': 1}`, Outbreak 2021 `metadata_summary = {'metadatafield2': 1}` and Outbreak 2022 `metadata_summary = {'metadatafield1': 1}`

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
